pkg/querier/queryrange/downstreamer.go

```diff
@@ -94,16 +99,50 @@ type instance struct {
 	parallelism int
 	locks       chan struct{}
 	handler     queryrangebase.Handler
+
+	splitAlign bool
+}
+
+// withoutOffset returns the given query string with offsets removed and timestamp adjusted accordingly. If no offset is present in original query, it will be returned as is.
+func withoutOffset(query logql.DownstreamQuery) (string, time.Time, time.Time) {
+	expr := query.Params.GetExpression()
+
+	var (
+		newStart = query.Params.Start()
+		newEnd   = query.Params.End()
+	)
+	expr.Walk(func(e syntax.Expr) {
+		switch rng := e.(type) {
+		case *syntax.RangeAggregationExpr:
+			off := rng.Left.Offset
+
+			if off != 0 {
+				rng.Left.Offset = 0     // remove offset
+				if newStart == newEnd { // instant query
```

Comment:
shouldn't we update the end irrespective of the range_type?

Reply:
also should we always expect a single offset in the expr? I am thinking of binary expressions that could use an offset on both operands. but [this code](https://github.com/grafana/loki/blob/85908fafdfc9261c2f12f630726fa0dad514f886/pkg/logql/rangemapper.go#L319) also expects a most 1 for some reason

Reply:
instant queries with binary expr that uses multiple offsets may be broken already

Reply:
> shouldn't we update the end irrespective of the range_type?

Good catch. you are right. End need to be updated irrespective of `range_type`. I think I added this in the very beginning to catch any instances if we use RangeMapper for range queries (turns out we use it only for instant queries).

Will update it 👍 

> also should we always expect a single offset in the expr? I am thinking of binary expressions that could use an offset on both operands. 
> instant queries with binary expr that uses multiple offsets may be broken already

I think [we split binary expressions into two expressions](https://github.com/grafana/loki/blob/kavirajk%2Fcache-instant-queries2/pkg/logql/rangemapper.go#L142) before mapping it into actual downstream `ConcatExpr` type. 

At this level, where we create and append downstream expressions, there cannot be two offsets in a single expression.